cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/ApplicationVerificationStage.java

```diff
@@ -211,27 +211,31 @@ private void verifyWorkflowNode(ApplicationSpecification appSpec, WorkflowSpecif
   }
 
   private void verifyWorkflowFork(ApplicationSpecification appSpec, WorkflowSpecification workflowSpec,
-                                  WorkflowNode node) {
+                                  WorkflowNode node, Set<String> nodeNames) {
     WorkflowForkNode forkNode = (WorkflowForkNode) node;
     Preconditions.checkNotNull(forkNode.getBranches(), String.format("Fork is added in the Workflow '%s' without" +
                                                                        " any branches", workflowSpec.getName()));
 
     for (List<WorkflowNode> branch : forkNode.getBranches()) {
-      verifyWorkflowBranch(appSpec, workflowSpec, branch);
+      verifyWorkflowNodeList(appSpec, workflowSpec, branch, nodeNames);
     }
   }
 
   private void verifyWorkflowCondition(ApplicationSpecification appSpec, WorkflowSpecification workflowSpec,
-                                       WorkflowNode node) {
+                                       WorkflowNode node, Set<String> nodeNames) {
     WorkflowConditionNode condition = (WorkflowConditionNode) node;
-    verifyWorkflowBranch(appSpec, workflowSpec, condition.getIfBranch());
-    verifyWorkflowBranch(appSpec, workflowSpec, condition.getElseBranch());
+    verifyWorkflowNodeList(appSpec, workflowSpec, condition.getIfBranch(), nodeNames);
+    verifyWorkflowNodeList(appSpec, workflowSpec, condition.getElseBranch(), nodeNames);
   }
 
-  private void verifyWorkflowBranch(ApplicationSpecification appSpec, WorkflowSpecification workflowSpec,
-                                    List<WorkflowNode> branch) {
-    for (WorkflowNode n : branch) {
-      verifyWorkflowNode(appSpec, workflowSpec, n);
+  private void verifyWorkflowNodeList(ApplicationSpecification appSpec, WorkflowSpecification workflowSpec,
+                                    List<WorkflowNode> nodeList, Set<String> nodeNames) {
+    for (WorkflowNode n : nodeList) {
+      if (nodeNames.contains(n.getNodeId())) {
+        throw new RuntimeException(String.format("Node with the name '%s' added multiple times.", n.getNodeId()));
```

Comment:
Should we throw a checked exception here? I think if there is a failure to deploy an app because of a conflicting unique name in the workflow definition, it would be better to throw a ``ConflictException`` or its subclass, so we can return an HTTP_CONFLICT (409).